libs/vkd3d/swapchain.c

```diff
@@ -2409,6 +2613,178 @@ static CONST_VTBL struct IDXGIVkSwapChainFactoryVtbl dxgi_vk_swap_chain_factory_
     dxgi_vk_swap_chain_factory_CreateSwapChain,
 };
 
+bool dxgi_vk_swap_chain_low_latency_enabled(struct dxgi_vk_swap_chain* chain)
+{
+    return chain->present.low_latency_state.mode;
+}
+
+void dxgi_vk_swap_chain_latency_sleep(struct dxgi_vk_swap_chain* chain)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = &chain->queue->device->vk_procs;
+    VkLatencySleepInfoNV latency_sleep_info;
+    VkSemaphoreWaitInfo sem_wait_info;
+
+    if (chain->present.low_latency_state.mode && chain->present.using_low_latency_present_mode)
```

Comment:
I'm confused about the interpretation of VkLatencySurfaceCapabilitiesNV.

The spec says that the specified present modes will be "optimized for use with low latency mode". There is no suggestion nor VUIDs that says other modes prevents the use of NvLL2.

The way I interpret the rational of "optimized" is that, latency optimization for render queue and latency optimization for present queue are different things. When NvLL2 is enabled, the input sampling --> GPU work done path will always target low-latency. However, the GPU work done -> Present done latency will only target low-latency for the modes in VkLatencySurfaceCapabilitiesNV. This makes sense because some modes have no particular ties to display refresh cycle, e.g. IMMEDIATE or MAILBOX.

I'm not sure what NV's implementation is, but so far I don't think the entire LL feature needs to be gated behind an optimized mode. On the other hand I think an optimized presentation mode can disable VKD3D's built-in presentation latency throttling which is based on present_wait.

Reply:
Actually, I think there is no need to disable the throttling either; when latency is kept low, it should never get throttled, and the throttling only acts as a failsafe measure when work is queuing up too rapidly or something is broken.

Hence, it's probably not necessary to change any behavior depending on the whether the mode is optimized.

Reply:
The mode check is removed so this can be resolved now.